File: backend/app/schemas/proposition.py
"""Schémas Pydantic v2 — Module Proposition de dates locataire (Sprint 4B).

Workflow back-and-forth bailleur ↔ locataire plafonné à 4 tours (cf
migration 0046 + proposition_service.py).

États possibles (cf CHECK constraint ck_dossiers_statut_proposition) :
  non_propose                    — état initial, locataire n'a rien envoyé
  propose_par_locataire          — locataire a proposé, bailleur doit répondre
  contre_propose_par_bailleur    — bailleur a contre-proposé, locataire répond
  accepte                        — accord conclu, date_accord fixée
  refuse                         — refus définitif, peut être reset par locataire

Doctrine §B.10 : pas de faux statut. Les CHECK constraints DB + Literal
Python double-verrouillent. Les requests ne tolèrent pas de champ optionnel
ambigu (les commentaires sont strictement bornés à 500 chars).
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

# ── Types Literal (source unique côté Pydantic) ──────────────────────────────

DureeEnvisagee = Literal["court", "moyen", "long", "indetermine"]
StatutProposition = Literal[
    "non_propose",
    "propose_par_locataire",
    "contre_propose_par_bailleur",
    "accepte",
    "refuse",
]
LastProposedBy = Literal["locataire", "bailleur"]

# Préférences locataire (4 axes, structurés en JSONB côté DB).
AnimauxPreference = Literal["oui", "non", "sous_conditions"]
FlexibiliteDate = Literal["rigide", "plus_moins_1_semaine", "plus_moins_1_mois"]
MeublePreference = Literal["meuble", "non_meuble", "indifferent"]


# ── Preferences ───────────────────────────────────────────────────────────────


class PreferencesLocataire(BaseModel):
    """Préférences structurées (persistées en JSONB sur dossiers_locataires)."""

    animaux: AnimauxPreference | None = None
    flexibilite_date: FlexibiliteDate | None = None
    colocation: bool | None = None
    meuble: MeublePreference | None = None


# ── Requests ──────────────────────────────────────────────────────────────────


class ProposerDatesRequest(BaseModel):
    """Body POST /locataires/{id}/proposition/proposer (locataire — état initial).

    Date d'entrée souhaitée OBLIGATOIRE — c'est le pivot du workflow. La durée
    et les préférences sont optionnelles mais fortement recommandées (UI les
    affiche en pré-remplissage côté bailleur).
    """

    date_entree_souhaitee: date
    duree_envisagee: DureeEnvisagee | None = None
    preferences: PreferencesLocataire = Field(default_factory=PreferencesLocataire)
    commentaire: str | None = Field(None, max_length=500)

    @field_validator("commentaire")
    @classmethod
    def _trim_or_none(cls, v: str | None) -> str | None:
        if v is None:
            return None
        trimmed = v.strip()
        return trimmed or None


class ContrePropositionBailleurRequest(BaseModel):
    """Body POST /locataires/{id}/proposition/contre-proposer (bailleur)."""

    date_contre_proposee: date
    commentaire: str | None = Field(None, max_length=500)

    @field_validator("commentaire")
    @classmethod
    def _trim_or_none(cls, v: str | None) -> str | None:
        if v is None:
            return None
        trimmed = v.strip()
        return trimmed or None


class ReContrePropositionLocataireRequest(BaseModel):
    """Body POST /locataires/{id}/proposition/re-contre-proposer (locataire).

    Identique à `ProposerDatesRequest` (la nouvelle date proposée) mais sans
    re-saisir préférences/durée déjà capturées au tour initial — seule la
    date change.
    """

    date_entree_souhaitee: date
    commentaire: str | None = Field(None, max_length=500)

    @field_validator("commentaire")
    @classmethod
    def _trim_or_none(cls, v: str | None) -> str | None:
        if v is None:
            return None
        trimmed = v.strip()
        return trimmed or None


class RefuserPropositionRequest(BaseModel):
    """Body POST /locataires/{id}/proposition/refuser (bailleur OU locataire)."""

    motif_refus: str | None = Field(None, max_length=500)

    @field_validator("motif_refus")
    @classmethod
    def _trim_or_none(cls, v: str | None) -> str | None:
        if v is None:
            return None
        trimmed = v.strip()
        return trimmed or None


# ── Response ──────────────────────────────────────────────────────────────────


class PropositionStatusResponse(BaseModel):
    """Réponse GET /locataires/{id}/proposition (status complet + flags UI).

    Les flags `peut_*` sont calculés côté backend et permettent au frontend
    d'afficher / cacher les CTA sans dupliquer la logique de transition.
    """

    model_config = ConfigDict(from_attributes=True)

    # État
    locataire_id: uuid.UUID
    statut_proposition: StatutProposition
    proposition_count: int

    # Données locataire
    date_entree_souhaitee: date | None = None
    duree_envisagee: DureeEnvisagee | None = None
    preferences: PreferencesLocataire = Field(default_factory=PreferencesLocataire)
    commentaire_locataire: str | None = None

    # Données bailleur
    date_contre_proposee_bailleur: date | None = None
    commentaire_bailleur: str | None = None

    # Final state
    motif_refus: str | None = None
    date_accord: date | None = None

    # Tracking
    last_proposed_at: datetime | None = None
    last_proposed_by: LastProposedBy | None = None

    # Flags UI (computed côté backend)
    peut_proposer: bool
    peut_contre_proposer: bool
    peut_accepter_locataire: bool
    peut_accepter_bailleur: bool
    peut_re_contre_proposer: bool
    peut_refuser: bool
    peut_reset: bool
    limite_atteinte: bool

File: frontend/src/lib/api/proposition.ts
"use client";

/**
 * API wrappers — Module Proposition de dates locataire (Sprint 4B).
 *
 * Endpoints backend : backend/app/routers/proposition.py.
 *
 * Workflow back-and-forth bailleur ↔ locataire plafonné à 4 tours :
 *   non_propose → propose_par_locataire → accepte / contre_propose_par_bailleur
 *                                           ↓
 *                                       refuse  or accepte / propose_par_locataire
 */

import { api } from "../api";

// ── Types ────────────────────────────────────────────────────────────────────

export type DureeEnvisagee = "court" | "moyen" | "long" | "indetermine";
export type StatutProposition =
  | "non_propose"
  | "propose_par_locataire"
  | "contre_propose_par_bailleur"
  | "accepte"
  | "refuse";
export type LastProposedBy = "locataire" | "bailleur";
export type AnimauxPreference = "oui" | "non" | "sous_conditions";
export type FlexibiliteDate = "rigide" | "plus_moins_1_semaine" | "plus_moins_1_mois";
export type MeublePreference = "meuble" | "non_meuble" | "indifferent";

export interface PreferencesLocataire {
  animaux?: AnimauxPreference | null;
  flexibilite_date?: FlexibiliteDate | null;
  colocation?: boolean | null;
  meuble?: MeublePreference | null;
}

export interface PropositionStatusResponse {
  locataire_id: string;
  statut_proposition: StatutProposition;
  proposition_count: number;

  date_entree_souhaitee: string | null;
  duree_envisagee: DureeEnvisagee | null;
  preferences: PreferencesLocataire;
  commentaire_locataire: string | null;

  date_contre_proposee_bailleur: string | null;
  commentaire_bailleur: string | null;

  motif_refus: string | null;
  date_accord: string | null;

  last_proposed_at: string | null;
  last_proposed_by: LastProposedBy | null;

  // Computed flags (UI affiche / masque les CTA en fonction)
  peut_proposer: boolean;
  peut_contre_proposer: boolean;
  peut_accepter_locataire: boolean;
  peut_accepter_bailleur: boolean;
  peut_re_contre_proposer: boolean;
  peut_refuser: boolean;
  peut_reset: boolean;
  limite_atteinte: boolean;
}

export interface ProposerDatesRequest {
  date_entree_souhaitee: string; // ISO date YYYY-MM-DD
  duree_envisagee?: DureeEnvisagee | null;
  preferences?: PreferencesLocataire;
  commentaire?: string | null;
}

export interface ContrePropositionBailleurRequest {
  date_contre_proposee: string;
  commentaire?: string | null;
}

export interface ReContrePropositionLocataireRequest {
  date_entree_souhaitee: string;
  commentaire?: string | null;
}

export interface RefuserPropositionRequest {
  motif_refus?: string | null;
}

// ── Labels (FR) ──────────────────────────────────────────────────────────────

export const DUREE_LABELS: Record<DureeEnvisagee, string> = {
  court: "Court terme (moins de 6 mois)",
  moyen: "Moyen terme (6 à 12 mois)",
  long: "Long terme (plus d'un an)",
  indetermine: "Durée indéterminée",
};

export const ANIMAUX_LABELS: Record<AnimauxPreference, string> = {
  oui: "Oui",
  non: "Non",
  sous_conditions: "Sous conditions",
};

export const FLEXIBILITE_LABELS: Record<FlexibiliteDate, string> = {
  rigide: "Date stricte",
  plus_moins_1_semaine: "± 1 semaine",
  plus_moins_1_mois: "± 1 mois",
};

export const MEUBLE_LABELS: Record<MeublePreference, string> = {
  meuble: "Meublé",
  non_meuble: "Non meublé",
  indifferent: "Indifférent",
};

export const MAX_PROPOSITIONS = 4;

// ── API calls ────────────────────────────────────────────────────────────────

export async function getProposition(
  locataireId: string,
): Promise<PropositionStatusResponse> {
  const { data } = await api.get<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition`,
  );
  return data;
}

export async function proposerDates(
  locataireId: string,
  payload: ProposerDatesRequest,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/proposer`,
    payload,
  );
  return data;
}

export async function contreProposerBailleur(
  locataireId: string,
  payload: ContrePropositionBailleurRequest,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/contre-proposer`,
    payload,
  );
  return data;
}

export async function accepterBailleur(
  locataireId: string,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/accepter-bailleur`,
    {},
  );
  return data;
}

export async function accepterLocataire(
  locataireId: string,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/accepter-locataire`,
    {},
  );
  return data;
}

export async function reContreProposerLocataire(
  locataireId: string,
  payload: ReContrePropositionLocataireRequest,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/re-contre-proposer`,
    payload,
  );
  return data;
}

export async function refuserProposition(
  locataireId: string,
  payload: RefuserPropositionRequest,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/refuser`,
    payload,
  );
  return data;
}

export async function resetProposition(
  locataireId: string,
): Promise<PropositionStatusResponse> {
  const { data } = await api.post<PropositionStatusResponse>(
    `/locataires/${locataireId}/proposition/reset`,
    {},
  );
  return data;
}
